### PR TITLE
TLS Inspector: Capture client hello size distribution.

### PR DESCRIPTION
**Pull request by KBaichoo**
Commit Message: TLS Inspector: Capture client hello size distribution.
Additional Description:
Risk Level: low
Testing: unit test
Docs Changes: included
Release Notes: included
Platform Specific Features: na
[Optional Runtime guard:] na
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc [hunk 1]
```diff
@@ -184,6 +204,9 @@ ParseState Filter::parseClientHello(const void* data, size_t len) {
     } else {
       config_->stats().tls_not_found_.inc();
     }
+    // Record client_hello size as we're done processing.
+    config_->stats().client_hello_size_.recordValue(
+        computeClientHelloSize(bio.get(), bytes_already_processed, len));
     return ParseState::Done;
   default:
     return ParseState::Error;
```
Comment (alyssawilk) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc:212]: should we track here?  how about having this and the above break out of the switch and consistently track bytes before returning ParseState::Error

File: docs/root/configuration/listeners/listener_filters/tls_inspector.rst [hunk 2]
```diff
@@ -35,15 +35,50 @@ Statistics
...
+
+  * - bytes_processed
+    - Histogram
+    - Records sizes which records the number of bytes the tls_inspector processed while analyzing for tls usage.
+      If the connection uses TLS: this is the size of client hello. If the client hello is too large, then the
+      recorded value will be 64KiB which is the maximum client hello size.
+      If the connection does not use TLS: it is the number of bytes processed
+      until the inspector determined the connection was not using TLS.
+      If the connection terminates early nothing is recorded if we didn't have
+      sufficient bytes for either of the cases above.
```
Comment (alyssawilk) [inline docs/root/configuration/listeners/listener_filters/tls_inspector.rst:82]: your call but we don't want to track bytes in progress for our analysis?

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc (hunk 1 above)
Comment (alyssawilk) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc:212]: I do think consistently tracking a catch all is slightly better coding hygiene though.  Any reason not to?

**Review by alyssawilk — COMMENTED** [review]
largely LGTM - tagging greg for cross-company TLS review

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc (hunk 1 above)
Comment (alyssawilk) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc:212]: I mean FWIW I agree looking at the ssl_read documentation there's _probably_ not other errors (WANT_WRITE is possible but unlikely?) but just as a code style thing we can have correctness rely on that being accurate and not changing, or we can track the stat outside the switch =P

File: docs/root/configuration/listeners/listener_filters/tls_inspector.rst (hunk 2 above)
Comment (KBaichoo) [reply docs/root/configuration/listeners/listener_filters/tls_inspector.rst:82]: I think modeling terminal states (no TLS after interpreting, TLS client hello detected / too large) makes sense for this histogram vs overloading with incomplete analysis.

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc (hunk 1 above)
Comment (KBaichoo) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc:212]: Done.

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review

File: test/extensions/filters/listener/tls_inspector/tls_inspector_test.cc [hunk 3]
```diff
@@ -276,6 +286,10 @@ TEST_P(TlsInspectorTest, ClientHelloTooBig) {
   auto state = filter_->onData(*buffer_);
   EXPECT_EQ(Network::FilterStatus::StopIteration, state);
   EXPECT_EQ(1, cfg_->stats().client_hello_too_large_.value());
+  const std::vector<uint64_t> bytes_processed =
+      store_.histogramValues("tls_inspector.bytes_processed", false);
+  ASSERT_EQ(1, bytes_processed.size());
+  EXPECT_EQ(50, bytes_processed[0]);
```
Comment (ggreenway) [inline test/extensions/filters/listener/tls_inspector/tls_inspector_test.cc:292]: For readability:

```suggestion
  EXPECT_EQ(max_size, bytes_processed[0]);
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
I reviewed this earlier but forgot to hit submit, and there's a comment on an earlier revision that I now can't figure out how to delete so ignore it.

lgtm except one small change in a test.

/wait

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc [hunk 4]
```diff
@@ -169,6 +186,9 @@ ParseState Filter::parseClientHello(const void* data, size_t len) {
       // We've hit the specified size limit. This is an unreasonably large ClientHello;
       // indicate failure.
       config_->stats().client_hello_too_large_.inc();
+      // Record client_hello size as we're done processing.
+      config_->stats().client_hello_size_.recordValue(
```
[the remark this answers is not in the document]
Comment (KBaichoo) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: Think this is "there's a comment on an earlier revision that I now can't figure out how to delete so ignore it."

Resolving.

File: test/extensions/filters/listener/tls_inspector/tls_inspector_test.cc (hunk 3 above)
Comment (KBaichoo) [reply test/extensions/filters/listener/tls_inspector/tls_inspector_test.cc:292]: done

**Review by ggreenway — APPROVED** [review]

**Pull request merged**